Enhanced `InferenceClient`

Wauplin opened the pull request:
**Goal:**

The goal of this PR is to refacto the existing InferenceAPI client. The aim is to:
1. **Support both InferenceAPI** (free) **and InferenceEndpoint** (paid product) in a single client
2. Provide a nice UX inspired by the [JS client](https://huggingface.co/docs/huggingface.js/inference/README#basic-examples) and namely:
   - **1 method per task.** E.g. `summary = client.summarization("this is a long text")`
   - list main parameters explicitly ([see js example](https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/tasks/nlp/summarization.ts#L5)). E.g `max_length`
   - parse output depending on the pipeline (text, image, audio,...)
   - recommend 1 model per task if user don't know which one to use
3. **Support custom requests** for Inference Endpoints (through a generic `.post()` method)

---

**Documentation:**
- [guide](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1474/en/guides/inference)
- [reference](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1474/en/package_reference/inference_api)
---

**Philosophy:** "easy to use is better than been optimal or exhaustive"

Examples:
- images parsed as `PIL.Image` by default
- inputs can be provided as bytes, files, URLs,...
- 1 recommended model per task
- only the main parameters are exposed (and let power users deal with more complex stuff) 
- `timeout` works for both "service unavailable" and proper timeout

---

```py
# Image classification with the default model
>>> from huggingface_hub import InferenceClient
>>> client = InferenceClient()
>>> client.image_classification("https://upload.wikimedia.org/wikipedia/commons/thumb/4/43/Cute_dog.jpg/320px-Cute_dog.jpg")
[{'score': 0.9779096841812134, 'label': 'Blenheim spaniel'}, ...]
```

---

**TODO (later)**
- [x] update documentation
- [ ] curate list of recommended models (+host it on the Hub)
- [ ] how to handle asyncio ? asyncio is much more relevant if we target devs
- [ ] validate inputs/outputs with pydantic ? => not a priority 
- [ ] deprecate `InferenceAPI`

osanseviero reviewed (commented):
Very cool! I [played with it](https://colab.research.google.com/drive/1GumOlzoR8hia7yZcEbRNEq1qIDOY6NBD?usp=sharing) and it seems to be working nicely!

Wauplin commented on src/huggingface_hub/_inference.py:
```diff
@@ -0,0 +1,100 @@
+
+
+def get_task_by_model_id(model_id: str) -> str:
+    raise NotImplementedError()
```
Thanks for letting me know! I didn't know about the second option.
I think I'll do option 1 to be explicit.

Narsil replied:
Option 2 will look for ModelInfo so... :)

Wauplin commented on src/huggingface_hub/_inference.py:
```diff
@@ -0,0 +1,100 @@
+
+    @staticmethod
+    def get_inference_api_url(model_id: str, task: str) -> str:
+        return f"{INFERENCE_ENDPOINT}/pipeline/{task}/{model_id}"
```
Would like a confirmation on this to be sure but this is the URL we are already using in `huggingface_hub` and it is the case since [the first PR](https://github.com/huggingface/huggingface_hub/pull/65/files#diff-f365bea0eb892dc99a45fae684d3df7562fd42011be7dedb3e41e35fed94f298R117) (2y ago).

osanseviero replied:
Yeah :thinking: I think we never officially communicated about this endpoint nor documented it as we pushed people to use the official one, but this one is more flexible and I think is ok to use here

Narsil replied:
Yes it's not officially documented as it causes confusion, and generally it's easier for everyone to keep 1 model - 1 task.

Biggest use case for this endpoint would be `feature-extraction` vs `sentence-similarity` (similarity is easier to understand in the widget form, but feature-extraction is the most useful one in a production setting where you're usually building a semantic database.

Wauplin replied:
Thanks for the details @Narsil. I think I'll keep the endpoint url with this form (i.e. /task/...). In the end it's just a url built for internal use so I can make the method private. Most users will not even realize it if they use the correct method for their model. 

And if a user tries to call a task on a model that doesn't support it, I'll gracefully handle the error to print to the user the available task(s) for their model.

Wauplin replied:
Realized after some experiments that if I can a wrong model <> task pair, the pipeline is loaded by Inference API and then an HTTP 500 is returned. Not so efficient to catch client-side (and server-side). Since this use case is mostly useful for feature-extraction task, I think I'll hardcode this one and use the more officially promoted `https://api-inference.huggingface.co/models/` URL in all other cases.

Narsil commented on src/huggingface_hub/_inference.py:
```diff
@@ -0,0 +1,100 @@
+#    https://huggingface.slack.com/archives/C03E4DQ9LAJ/p1680169099087869
+
+# TODO:
+# - handle options? wait_for_model, use_gpu,... See list: https://github.com/huggingface/huggingface.js/blob/main/packages/inference/src/types.ts#L1
```
Indeed. Right now only legacy membership have access to using GPUs. (We do use GPUs for some models, but we choose on which hardware to execute what now)

Wauplin replied:
Thinking twice about it, I'll not expose `wait_for_model` as an option for the user. One can set `InferenceClient(..., timeout=...)` to control how much time they want to wait (defaults to `timeout=None` meaning we'll wait). I'll still use `wait_for_model` internally even if not exposed.

I've found mentions of other options ([here](https://github.com/huggingface/huggingface.js/blob/b18832d90d3c7c43edbf72cfe976ba02360a4a5a/packages/inference/src/lib/makeRequestOptions.ts#LL33C16-L33C25) and [here](https://github.com/huggingface/huggingface.js/blob/b18832d90d3c7c43edbf72cfe976ba02360a4a5a/packages/inference/src/types.ts)) like `use_cache` and `dont_load_model` but I'll not expose them either.

Wauplin replied:
So basically to keep it simple, only `timeout: Optional[int] = None` will be exposed  (and leverage `wait_for_model` + handle HTTP 503 errors under the hood). Does that make sense?

Narsil commented on src/huggingface_hub/_inference.py:
```diff
@@ -0,0 +1,100 @@
+
+    def summarization(
+        self,
+        text: str,
```
I'd keep things simple if it's new code :)

`transformers` has lots of legacy in its acceptance of inputs and outputs.

Wauplin commented:
I started to add tasks but still wondering how I'll document/test/maintain this. Hopefully low maintenance once every task is implemented.

Here is a working example with current version. Audio and image in input can be provided as bytes, file path or url. In output, audio is converted to bytes while images are parsed as `PIL.Image`. The philosophy is _"let's make it as easy as possible for the user and it it's too restrictive, let it use the generic `client.post(...)`"_.


```py
>>> from huggingface_hub import InferenceClient
>>> client = InferenceClient()

# Text to speech to text
>>> audio = client.text_to_speech("Hello world")
b'fLaC\x00\x00...'
>>> client.audio_classification(audio)
[{'score': 0.4976358711719513, 'label': 'hap'}, {'score': 0.3677836060523987, 'label': 'neu'}, {'score': 0.1274358034133911, 'label': 'ang'}, {'score': 0.007144733797758818, 'label': 'sad'}]
>>> client.automatic_speech_recognition(audio)
'LOW WHIRLD'

# Image classification
>>> client.image_classification("cat.jpg")
[{'score': 0.41461074352264404, 'label': 'tabby, tabby cat'}, ...]
>>> client.image_classification("https://upload.wikimedia.org/wikipedia/commons/thumb/4/43/Cute_dog.jpg/320px-Cute_dog.jpg")
[{'score': 0.9779096841812134, 'label': 'Blenheim spaniel'}, ...]

# Image segmentation
>>> client.image_segmentation("cat.jpg"):
[{'score': 0.989008, 'label': 'LABEL_184', 'mask': <PIL.PngImagePlugin.PngImageFile image mode=L size=400x300 at 0x7FDD2B129CC0>}, ...]

# Text summarization
>>> client.summarization("The Eiffel tower...")
'The Eiffel tower is one of the most famous landmarks in the world....'

# Chat
>>> client.conversational("Hi, who are you?")
{'generated_text': 'I am the one who knocks.', 'conversation': {'generated_responses': ['I am the one who knocks.'], 'past_user_inputs': ['Hi, who are you?']}, 'warnings': ['Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.']}
>>> client.conversational(
...     "Wow, that's scary!",
...     generated_responses=output["conversation"]["generated_responses"],
...     past_user_inputs=output["conversation"]["past_user_inputs"],
... )
```

Wauplin commented:
List of tasks:
  audio
    - [x] audio classification
    - [x] speech to text (automatic_speech_recognition)
    - [x] text to speech
- computer vision
    - [x] image classification
    - [x] image segmentation
    - [x]  image to image
    - [x] image to text
    - [ ] object detection
    - [x] text to image
- multimodal
    - [ ] documentation question answering
    - [ ] visual question answering
- nlp
    - [x] conversational
    - [x] feature extraction
    - [ ] fill mask
    - [ ] question answering
    - [x] sentence similarity
    - [x] summarization
    - [ ] table question answering
    - [ ] text classifcation
    - [ ] **text generation** (started [here](https://github.com/huggingface/huggingface_hub/compare/enhanced-inference-client...enhanced-inference-client-text-generation?expand=1) but let's merge without it)
    - [ ] token classification
    - [ ] translation
    - [ ] zero shot classification
- [ ] tabular
    - [ ] tabular classification
    - [ ] tabular regression

osanseviero commented on src/huggingface_hub/_inference.py:
```diff
@@ -0,0 +1,886 @@
+                audio file.
+            model (`str`, *optional*):
+                The model to use for audio classification. Can be a model ID hosted on the Hugging Face Hub
+                or a URL to a deployed Inference Endpoint. If not provided, the default recommended model for
```
I would mention the default here

osanseviero commented on src/huggingface_hub/_inference.py:
```diff
@@ -0,0 +1,886 @@
+    # TODO: load from a config file? (from the Hub?) Would make sense to make updates easier.
+    if task in RECOMMENDED_MODELS:
+        model = RECOMMENDED_MODELS[task]
+        logger.info(f"Defaulting to recommended model {model} for task {task}.")
```
Maybe mention here that users can specify model using `model` param

Wauplin replied:
Good idea. Also mentioning that we do not guarantee backward compatibility

osanseviero reviewed (commented):
Thanks a lot! I did a first pass

stevhliu commented on docs/source/guides/inference.mdx:
```diff
@@ -1,58 +1,314 @@
+
+## Custom requests
+
+Despite the simplicity of the [`InferenceClient`] API, it is not always possible to cover all use cases. The
+[`InferenceClient.post`] method allows you to send any request to the Inference API. With this flexibility, you can
+for example handle yourself the parsing of the inputs and output. 
```
```suggestion
However, it is not always possible to cover all use cases. For custom requests, the
[`InferenceClient.post`] method gives you the flexibility to send any request to the Inference API. For example, you can specify how to parse the inputs and outputs:
```

stevhliu commented on docs/source/guides/inference.mdx:
```diff
@@ -1,58 +1,314 @@
 
-The metadata in the model card and configuration files (see [here](https://huggingface.co/docs/hub/models-widgets#enabling-a-widget) for more details) determines the pipeline type. For example, when using the [bert-base-uncased](https://huggingface.co/bert-base-uncased) model, the Inference API can automatically infer that this model should be used for a `fill-mask` task.
+In the example above, the generated image is returned as raw bytes. This can be helpful if you don't have `Pillow`
+installed in your setup and just care about the binary content of the image. [`InferenceClient.post`] is also useful
+to handle tasks that are not yet officially supported.
```
Maybe move this to the intro of the Custom requests section. Feels a bit off to introduce/mention this at the end.

stevhliu commented on docs/source/guides/inference.mdx:
```diff
@@ -1,58 +1,314 @@
+
+### Timeout
+
+When doing inference, there are two main causes for a timeout to happen:
```
```suggestion
When doing inference, there are two main causes for a timeout:
```

stevhliu commented on docs/source/guides/inference.mdx:
```diff
@@ -1,58 +1,314 @@
-If you want to make the HTTP calls directly, please refer to [Accelerated Inference API Documentation](https://api-inference.huggingface.co/docs/python/html/index.html) or to the sample snippets visible on every supported model page.
+[`InferenceClient`] acts as a replacement for the legacy [`InferenceApi`] client. It adds specific support for tasks and
+handles inference on both [Inference API](https://huggingface.co/docs/api-inference/index) and [Inference Endpoints](https://huggingface.co/docs/inference-endpoints/index).
+Learn how to migrate to the new client in this section.
```
```suggestion
Learn how to migrate to the new client in the[Legacy InferenceAPI client](#legacy-inferenceapi-client) section.
```

stevhliu commented on docs/source/package_reference/inference_client.mdx:
```diff
@@ -0,0 +1,48 @@
+Inference is the process of using a trained model to make predictions on new data. This process can be compute-intensive
+which makes it interesting to run on a remote machine. The `huggingface_hub` library provides an easy way to call a
+service that runs fast inference for hosted models. There are several services you can connect to:
+- [Inference API](https://huggingface.co/docs/api-inference/index): a service that allows you to run accelerated inference
```
Same comments for this section as the ones I left on `inference_api.mdx` :)

stevhliu commented on docs/source/package_reference/inference_client.mdx:
```diff
@@ -0,0 +1,48 @@
+
+## InferenceAPI
+
+[`InferenceAPI`] is the legacy way to call the Inference API. The interface is more simplistic and requires to know
+the input parameters and output format for each task. It also lacks the ability to connect to other services like
```
```suggestion
[`InferenceAPI`] is the legacy way to call the Inference API. The interface is more simplistic and requires knowing
the input parameters and output format for each task. It also lacks the ability to connect to other services like
```

stevhliu commented on src/huggingface_hub/_inference.py:
```diff
@@ -0,0 +1,886 @@
+    Args:
+        model (`str`, `optional`):
+            The model to run inference with. Can be a model id hosted on the Hugging Face Hub, e.g.
+            `bigcode/starcoder` or a URL to a deployed Inference Endpoint. Defaults to None, meaning the model can
+                be passed for each task or in last resort a recommended model will be used.
```
```suggestion
            `bigcode/starcoder` or a URL to a deployed Inference Endpoint. Defaults to None, in which case a recommended model is automatically selected for the task.
```

stevhliu commented on src/huggingface_hub/_inference.py:
```diff
@@ -0,0 +1,886 @@
+            generated_responses (`List[str]`, *optional*):
+                A list of strings corresponding to the earlier replies from the model. Defaults to None.
+            past_user_inputs (`List[str]`, *optional*):
+                A list of strings corresponding to the earlier replies from the user. Should be of the same length of
+                `generated_responses`. Defaults to None.
```
```suggestion
                A list of strings corresponding to the earlier replies from the user. Should be the same length as
                `generated_responses`. Defaults to None.
```

stevhliu commented on src/huggingface_hub/_inference.py:
```diff
@@ -0,0 +1,886 @@
+        """
+        Perform image segmentation on the given image using the specified model.
+
+        <Tip warning={true}>
```
Should we also include this warning in the `image_classification` docstring?

Wauplin replied:
Actually no because `image_classification` can work simply with an image file as input (and no image as output). In the case of image_segmentation the output contains an image which makes it mandatory.

stevhliu commented on src/huggingface_hub/_inference.py:
```diff
@@ -0,0 +1,886 @@
+
+        Args:
+            image (`Union[str, Path, bytes, BinaryIO]`):
+                The input image for translation. It can be raw bytes, an image file, or a URL to an online image..
```
```suggestion
                The input image for translation. It can be raw bytes, an image file, or a URL to an online image.
```

stevhliu commented on src/huggingface_hub/_inference.py:
```diff
@@ -0,0 +1,886 @@
+        """
+        Takes an input image and return text.
+
+        Models can be very different outputs depending on your use case (image captioning, optical character recognition
```
```suggestion
        Models can have very different outputs depending on your use case (image captioning, optical character recognition
```

stevhliu commented on src/huggingface_hub/_inference_types.py:
```diff
@@ -0,0 +1,70 @@
+
+
+class ClassificationOutput(TypedDict):
+    """Dictionary containing the output of a 'classification' task.
```
Might be nice to be able to link back to the specific task :)

```suggestion
    """Dictionary containing the output of a [`~InferenceClient.audio_classification`] and  [`~InferenceClient.image_classification`] task.
```

stevhliu reviewed (approved):
Very cool guide and comprehensive API documentation! 💯

Love it when you ping me to look at something because I always learn something new!

Wauplin commented on docs/source/guides/inference.mdx:
```diff
@@ -1,58 +1,314 @@
->>> from huggingface_hub.inference_api import InferenceApi
->>> inference = InferenceApi(repo_id="bert-base-uncased", token=API_TOKEN)
+>>> from huggingface_hub import InferenceClient
+>>> client = InferenceClient()
```
I would rather not add a token parameter. I feel that if we add it everywhere we encourage users to pass tokens as argument instead of log in correctly.

Wauplin commented on src/huggingface_hub/_inference_types.py:
```diff
@@ -0,0 +1,70 @@
+
+
+class ConversationalOutputConversation(TypedDict):
+    """Dictionary containing the "conversation" part of a 'conversational' task.
```
```suggestion
    """Dictionary containing the "conversation" part of a [`~InferenceClient.conversational`] task.
```

Wauplin commented on src/huggingface_hub/_inference_types.py:
```diff
@@ -0,0 +1,70 @@
+
+
+class ConversationalOutput(TypedDict):
+    """Dictionary containing the output of a 'conversational' task.
```
```suggestion
    """Dictionary containing the output of a  [`~InferenceClient.conversational`] task.
```

Wauplin commented on src/huggingface_hub/_inference_types.py:
```diff
@@ -0,0 +1,70 @@
+
+
+class ImageSegmentationOutput(TypedDict):
+    """Dictionary containing information about a 'image-segmentation' task. In practice, image segmentation returns a
```
```suggestion
    """Dictionary containing information about a [`~InferenceClient.image_segmentation`] task. In practice, image segmentation returns a
```

Wauplin commented on docs/source/guides/inference.mdx:
```diff
@@ -1,58 +1,314 @@
+[Inference Endpoints](https://huggingface.co/docs/inference-endpoints/index) documentation pages.
+
+For web development, a [JS client](https://huggingface.co/docs/huggingface.js/inference/README) has been released.
+If you are interested in game development, you might have a look to our [C# project](https://github.com/huggingface/unity-api).
```
That's what I meant just above with _If you want to make the HTTP calls directly, please
refer to the Inference API or Inference Endpoints doc pages_

Wauplin replied:
I'll add a "using the tool of your choice (curl, postman,...)"

Wauplin commented on src/huggingface_hub/_inference_types.py:
```diff
@@ -0,0 +1,70 @@
+if TYPE_CHECKING:
+    from PIL import Image
+
+
```
To be honest, this module is mostly there to help IDEs and users with type annotations. We don't do any kind of input/output validation for now (might be the case in the future)

Wauplin commented on src/huggingface_hub/_inference.py:
```diff
@@ -0,0 +1,886 @@
+        )
+
+
+def _get_recommended_model(task: str) -> str:
```
No I don't want to guarantee backward compatibility. I would rather mention in the docs + in the log message that we recommend passing explicitly the model instead. As I see it, we should aim to curate the best model for our users. The _"best"_ model being one with good performance but also that we serve efficiently in Inference API. This is why I think it should be de-correlated from the `transformers` (and `diffusers`) pipelines.

Also I'm in favor of hosting a json file on the Hub to store the recommended models. This way we could update the list without making a release of `huggingface_hub`. The same list could also be shared with the JS client as well.
=> this is also why I'd rather not document the recommended model in the docstrings.

Wauplin commented on src/huggingface_hub/_inference.py:
```diff
@@ -0,0 +1,886 @@
+
+
+RECOMMENDED_MODELS = {
+    "audio-classification": "superb/hubert-large-superb-er",
```
See https://github.com/huggingface/huggingface_hub/pull/1474#discussion_r1210402514 about why not aligning with `transformers`. I'm not opinionated about which recommendations to do as long as they are well served by Inference API.

Wauplin commented on src/huggingface_hub/_inference.py:
```diff
@@ -0,0 +1,886 @@
+logger = logging.getLogger(__name__)
+
+
+RECOMMENDED_MODELS = {
```
See https://github.com/huggingface/huggingface_hub/pull/1474#discussion_r1210402514.

Wauplin commented on docs/source/guides/inference.mdx:
```diff
@@ -1,58 +1,314 @@
+[`InferenceClient`]'s goal is to provide the most easy-to-use interface to run inference on Hugging Face models. It
+has a simple API that supports the most common tasks. Here is a list of the currently supported tasks:
+
+- audio
```
Good idea. I updated it in https://github.com/huggingface/huggingface_hub/pull/1474/commits/978fbf29ea7107772d9484f0d47fa4572aab8fad. I removed the "models" columns with number of models since it would need to be manually updated and I'm pretty sure it will not :smile: Having a link to the hf.co/tasks page is really nice though.

Wauplin commented:
@osanseviero @stevhliu Thanks a TON for the thorough review :pray: I think I addressed all the points.
The main remaining question was about recommended models (see https://github.com/huggingface/huggingface_hub/pull/1474#discussion_r1210402514). The TL;DR: I'd rather not guarantee anything and will later default to have an hosted config file for them.

osanseviero reviewed (approved):
Looks good!

Wauplin commented:
I finally merged this one! :) And opened a followup issue to list the next steps: https://github.com/huggingface/huggingface_hub/issues/1488
